ios/keyman/Keyman/Keyman/AppDelegate.swift

```diff
@@ -91,6 +113,74 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     return _overlayWindow!
   }
 
+  public func installAdhocKeyboard(url: URL) {
+    let documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
+    var destination =  documentsDirectory
+    destination.appendPathComponent("temp/\(url.lastPathComponent)")
+
+    Manager.shared.unzipFile(fileUrl: url, destination: destination) {
+      self.promptAdHocInstall(destination)
+    }
+  }
+
+  public func promptAdHocInstall(_ folder: URL) {
+    _adhocDirectory = folder
+
+    var welcomePath = folder
+    welcomePath.appendPathComponent("Welcome.htm")
+
+    let vc = UIViewController()
+    vc.view.backgroundColor = .red
+    let wkWebView = WKWebView.init(frame: vc.view.frame)
+    wkWebView.backgroundColor = .blue
+    vc.view.addSubview(wkWebView)
+    let cancelBtn = UIBarButtonItem(title: "Cancel", style: .plain,
+                                    target: self,
+                                    action: #selector(cancelAdHocBtnHandler))
+    let installBtn = UIBarButtonItem(title: "Install", style: .plain,
+                                     target: self,
+                                     action: #selector(installAdHocBtnHandler))
+    vc.navigationItem.leftBarButtonItem = cancelBtn
+    vc.navigationItem.rightBarButtonItem = installBtn
+    let nvc = UINavigationController.init(rootViewController: vc)
+
+    self.window?.rootViewController?.present(nvc, animated: true, completion: {
+      if FileManager.default.fileExists(atPath: welcomePath.path) {
+        print("loading request: \(welcomePath)")
+        if let html = try? String(contentsOfFile: welcomePath.path, encoding: String.Encoding.utf8) {
+          wkWebView.loadHTMLString(html, baseURL: nil)
+        } else {
+           wkWebView.loadHTMLString("Keyboard Package", baseURL: nil)
+        }
+      } else {
+        wkWebView.loadHTMLString("Keyboard Package", baseURL: nil)
```

Comment:
If welcomePath doesn't exist or can't be loaded, we should display the package ID and package version here